type-definitions/index.d.ts

```diff
@@ -0,0 +1,20 @@
+import {Stream} from "most";
+
+// all 3 args
+export function sample<A, B, C>(f: (a: A, b: B) => C, sampler: Stream<A>, s: Stream<B>): Stream<C>;
+// first arg then 2
+export function sample<A, B, C>(f: (a: A, b: B) => C): (sampler: Stream<A>, s: Stream<B>) => Stream<C>;
+// first 2 then 1
+export function sample<A, B, C>(f: (a: A, b: B) => C, sampler: Stream<A>): (s: Stream<B>) => Stream<C>;
+// 1 arg at a time
+export function sample<A, B, C>(f: (a: A, b: B) => C): (sampler: Stream<A>) => (s: Stream<B>) => Stream<C>;
+
+
+// all 3 args
+export function sampleArray<A, B, C>(f: (a: A, b: B[]) => C, sampler: Stream<A>, streams: Stream<B>[]): Stream<C>;
+// first arg then 2
+export function sampleArray<A, B, C>(f: (a: A, b: B[]) => C): (sampler: Stream<A>, s: Stream<B>[]) => Stream<C>;
+// first 2 then 1
+export function sampleArray<A, B, C>(f: (a: A, b: B[]) => C, sampler: Stream<A>): (s: Stream<B>[]) => Stream<C>;
+// 1 arg at a time
+export function sampleArray<A, B, C>(f: (a: A, b: B[]) => C): (sampler: Stream<A>) => (s: Stream<B>[]) => Stream<C>;
```

Comment:
This is pretty brutal already, but I think it may need to be expanded even more if we want to really do it right.  Like [combineArray](https://github.com/cujojs/most/blob/master/type-definitions/most.d.ts#L122), I think we need to add polymorphic variants of sampleArray.  Unfortunately, having to handle both partial application *and* polymorphism will cause a combinatoric explosion of type definitions :(  Any ideas on other ways to do it?